compiler/rustc_target/src/abi/mod.rs

```diff
@@ -1162,8 +1162,13 @@ impl<'a, Ty> TyAndLayout<'a, Ty> {
         // If we have not found an error yet, we need to recursively descend into fields.
         match &self.fields {
             FieldsShape::Primitive | FieldsShape::Union { .. } => {}
-            FieldsShape::Array { .. } => {
-                // FIXME(#66151): For now, we are conservative and do not check arrays.
+            FieldsShape::Array { count, .. } => {
+                if !(*count == 0
+                    || self.field(cx, 0).to_result()?.might_permit_raw_init(cx, zero)?)
```

Comment:
Can you avoid nesting a disjunction inside `!`? I think this would be more readable:
`count > 0 && !self.field....`

Reply:
Fixed